staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go

```diff
@@ -29,51 +29,98 @@ import (
 	"k8s.io/client-go/pkg/api/v1"
 )
 
-// ResourceLocation returns a URL to which one can send traffic for the specified service.
-func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string) (*url.URL, error) {
+func parseID(id string) (scheme, name, port string, err error) {
+	if len(id) == 0 {
+		return "", "", "", fmt.Errorf("invalid empty service name %q", id)
+	}
+
 	// Allow ID as "svcname", "svcname:port", or "scheme:svcname:port".
 	svcScheme, svcName, portStr, valid := utilnet.SplitSchemeNamePort(id)
 	if !valid {
-		return nil, errors.NewBadRequest(fmt.Sprintf("invalid service request %q", id))
+		return "", "", "", errors.NewBadRequest(fmt.Sprintf("invalid service request %q", id))
+	}
+
+	if len(svcScheme) == 0 {
+		svcScheme = "https"
+	}
+	if len(portStr) == 0 {
+		switch svcScheme {
+		case "http":
+			portStr = "80"
+		default:
+			portStr = "443"
+		}
+	}
+	return svcScheme, svcName, portStr, nil
+}
+
+// resolveToService returns the service, scheme and the port string (can be a name, can be a numeric port).
+func resolveToService(services listersv1.ServiceLister, namespace, id string) (*v1.Service, string, string, error) {
+	svc, err := services.Services(namespace).Get(id)
+	if err != nil {
+		return nil, "", "", err
 	}
 
+	return svc, "https", "443", nil
```

Comment:
Updated the code here from #45082, i.e. the caller decides.

The HTTP removal is intentional, as discussed in #45082 with @deads2k. For dev and testing one can disable cert verification and use self-signed certs. So we don't need HTTP anymore.